app/models/miq_region.rb

```diff
@@ -242,6 +243,45 @@ def remote_ws_url
     hostname.nil? ? nil : "https://#{hostname}"
   end
 
+  def generate_auth_key(ssh_user, ssh_password, ssh_host = remote_ws_address)
+    key = remote_region_v2_key(ssh_user, MiqPassword.try_decrypt(ssh_password), ssh_host)
+
+    auth = AuthToken.new
+    auth.auth_key = key
+    auth.name = "Region #{region} API Key"
+    auth.resource = self
+    auth.save!
+  end
+
+  def verify_credentials(_auth_type = nil, _options = nil)
+    # TODO: verify the key against the remote api using the api client gem
+    true
+  end
+
+  def api_system_auth_token(userid)
+    region_v2_key = authentication_token
+
+    token_hash = {
+      :server_guid => remote_ws_miq_server.guid,
+      :userid      => userid,
+      :timestamp   => Time.now.utc
+    }
```

Comment:
this is fine, though as a separate discussion, let's review if we need better validation of server_guid on the API server side (and probably updating here accordingly).